00xx-account-deletion.md

```diff
@@ -0,0 +1,576 @@
+this marks the empty account.
+
+
+```
+    H(H(A,B),H(C,x))               free = []
+      /       \
+   H(A,B)     H(C,x)
+   /  \       /  \
+  A    B     C    x
+```
+
+The free list contains a set of locations bound to the empty account located
+between the leftmost leaf and righmost non-empty leaf.  The idea is to recycle
+these locations upon further addition of data, as we will see shortly.
+
+
+The insertion of new data `D` results in the following Merkle tree.
+
+```
+    H(H(A,B),H(C,D))               free = []
+      /       \
+   H(A,B)     H(C,D)
+   /  \       /  \
+  A    B     C    D
+```
+
+Upon removal of `C`, the structure would evolve as follows, with `x` (the empty
+account) marking the freed location.
+
+In this example, locations will be represented as lists of directions for the
+sake of readability. In practice, though, they will be represented by other,
+usually isomorphic to the list of directions, means, e.g., an integer encoding
+the position of the leaf.
+
+The free list now states that the location determined by
+sequence of directions `[Right; Left]` is available for reuse for new
+insertions.
+
+```
+    H(H(A,B),H(C,D))       free = [[Right; Left]]
+      /       \
+   H(A,B)     H(x,D)
+   /  \       /  \
+  A    B     x    D
+
+```
+
+Thus removal incurs a cost proportional to the height of the tree since we have
+to recompute the hashes on the Merkle path from the removed data.
+
+Now adding data `E` would result in
+
+```
+    H(H(A,B),H(E,D))               free = []
+      /       \
+   H(A,B)     H(E,D)
+   /  \       /  \
+  A    B     E    D
+```
+
+##### <a name="free_list repr"></a>Representation of the free list
+
+The view of the free list shown in the example above is a simplification.
+
+The contents of this list need be easily shareable, e.g., to synchronize a
+bootstrapping node. In order to achieve that, the data structure representing
+the free list needs to be *ordered*.
+
+The ordering can be derived from the timestamp the locations are freed during
+the lifetime of the chain. This would result in using a *stack* (for a LIFO
+allocation) or a *queue* (for a FIFO allocation). However, this choice has some
+drawbacks (see [alternative](#free_list_repr alt)), most importantly that a
+newcomer node cannot have any knowledge of the past, so that the free list would
+need to be fully communicated to it.
+
+**Imposing an ordering** from the structure of the data that we insert in the
+free list has better properties. Heaps, `OCaml` sets (which rely on a comparison
+function), or even ordered lists are implementation choices that would work in
+this context.
+
+We thus propose to use a **max-heap** (like
+[bheap](https://opam.ocaml.org/packages/bheap/) or
+[CCHeap](https://c-cube.github.io/ocaml-containers/last/containers/CCHeap/index.html))
+for this implementation, that is, the free list is an ordered set of locations,
+from which we always allocate the biggest one.
```

Comment:
Why are we choosing 'biggest' here? My intuition is that, especially if this feature is widely used, we might have accounts created and deleted at the right side of the tree; it would be natural to treat this as moving the fill frontier backwards. For example, if we had a series of deletes like
```
[a, b, c, d, e, f]
[a, b, c, X, e, f]
[a, b, c, X, e, X]
[a, b, c, X, X, X]
```
then it's natural to treat the states as
```
[a, b, c, d, e, f]
[a, b, c, X, e, f]
[a, b, c, X, e]
[a, b, c]
```
and recover the memory. However, in the current description, an insertion would give us
```
[a, b, c, X, X, g]
```
where in a sense we're always trying to 'waste' the maximum possible number of spaces by choosing the largest. If we instead choose the smallest, it really does behave like we've reduced the fill frontier, and always gives us the option to do that as a performance optimisation.

Reply:
I explain this in the example right below in the document. There are 2 elements for allocations in the structure:
1. The free list (indices between [0, fill frontier index[), let's call it l
2. The fill frontier index, let's call it i

The strategy would be to always use the free list first (from the biggest to the lowest index), then the fill frontier index if the the free list is empty. 

Replaying your example, this would give 
```
[a, b, c, d, e, f]       l = [] i = 5
[a, b, c, X, e, f]      l = [3]  i = 5
[a, b, c, X, e, X]     l = [3] i = 4
[a, b, c, X, X, X]     l = []  i = 2 ( this is where it's important to have a max heap, to make this reduction easy, since you just have to peek and pop)
``
Thus the next insertion would give 
`[ a, b, c, g]`as you described since the free list has been emptied.

Reply:
There are 2 things in conflict here: coalescing the free space to the right while also inserting to the left to keep it dense. You've addressed one, I was addressing the other. Please talk to both in the RFC. It seems as though either option probably require us to deserialise the whole list (unless we have head and bottom peeking).

Reply:
Creating then deleting an account at the fill frontier is already correctly handled [here](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_mask/masking_merkle_tree.ml#L497) by just moving the `current_location` (what I call the fill frontier).  

I'm not sure what I am missing. Could you show me another example of where what I am proposing doesn't offer a good solution.

Reply:
```
[a, b, c, X, X, f, X]     l = [4, 3] i = 4
[a, b, c, X, e, f, X]     l = [3]  i = 2
```
This is bad.

Reply:
No in your example `i` would still be 4. Allocating from the free list does not change the fill frontier.

Reply:
In reverse, deallocation from the fill frontier might affect the free list, like in the example above

Reply:
I will add a section on allocation strategy to make that clearer. It does not seem to be.

Reply:
The contents of this discussion is reflected in the document